internal/types/subtypes/attribute_transform.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	stAttrField, err := attributeField(d, st)
+	if err != nil {
+		// no need to convert
+		return nil
+	}
+
+	if defaultAttrField == stAttrField {
+		// no need to convert
+		return nil
+	}
+
+	stAttrs, ok := r.Get(stAttrField).Message().Interface().(proto.Message)
+	if !ok {
+		return fmt.Errorf("found subtype attribute field that is not proto.Message: %s %s", d.FullName(), stAttrField.FullName())
+	}
+	defaultAttrs, err := protoToStruct(stAttrs)
+	if err != nil {
+		return err
+	}
+	// implicitly clears any previously set oneof value
+	r.Set(defaultAttrField, protoreflect.ValueOfMessage(defaultAttrs.ProtoReflect()))
+	return nil
+}
+
+func structToProto(fields *structpb.Struct, p proto.Message) error {
+	if fields == nil {
+		// If there is no struct, don't update the default proto message.
+		return nil
+	}
+	js, err := fields.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	// TODO: replicate this logic but with a proto extension set on the Field
+	// descriptor There are some attributes where we want to discard unknown
+	// fields, while others that should error if there are unknown fields
+	//
+	// opts := GetOpts(opt...)
+	// if opts.withDiscardUnknownFields {
+	// 	err = (protojson.UnmarshalOptions{DiscardUnknown: true}.Unmarshal(js, p))
+	// } else {
+	// 	err = protojson.Unmarshal(js, p)
+	// }
+
+	err = protojson.Unmarshal(js, p)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func protoToStruct(p proto.Message) (*structpb.Struct, error) {
+	js, err := protojson.MarshalOptions{UseProtoNames: true}.Marshal(p)
+	if err != nil {
+		return nil, err
+	}
+	st := &structpb.Struct{}
+	if err := protojson.Unmarshal(js, st); err != nil {
+		return nil, err
+	}
+	return st, nil
+}
+
+// Filterable converts a proto.Message so any subtype attributes fields are
+// structed like the API so filter strings will be correctly applied. If the
+// given proto.Message does not have any subtype attributes, the original
+// proto.Message is returned.  To determine if the message needs
+// transformation, it looks for a oneof field named "attrs". It also expects
+// that there is a structpb.Struct field named "attributes" as part of the
+// oneof. Thus the message must be like:
+//
+//    message Foo {
+//      // other fields
+//      oneof attrs {
+//        google.protobuf.Struct attributes = 100;
+//        // other attribute fields
+//      }
+//    }
+//
+// If the message does not conform to this structure,
+// the original message is returned.
+func Filterable(item proto.Message) (proto.Message, error) {
+	clone := proto.Clone(item)
+	r := clone.ProtoReflect()
+
+	attrsField := r.Descriptor().Oneofs().ByName("attrs")
+	if attrsField == nil {
+		return item, nil
+	}
+
+	defaultAttrField := attrsField.Fields().ByName("attributes")
+	if defaultAttrField == nil {
+		return item, nil
+	}
+
+	var attr proto.Message
+	var pbAttrs proto.Message
+	var err error
+
+	oneOfFields := attrsField.Fields()
+
+	for i := 0; i < oneOfFields.Len(); i++ {
```

Comment:
Would it be worth adding a comment here about what we're doing? The only clue is the `if` statement, so maybe worth spelling it out.
```suggestion
	// Find the populated oneof field and turn it into a `*structpb.Struct`.
	for i := 0; i < oneOfFields.Len(); i++ {
```